feat(cart): Implement and standardize Cart API endpoints

irvaniamirali opened the pull request:
Implemented all Cart endpoints (`POST /cart/`, `GET /cart/{user_id}`, `DELETE /cart/{user_id}`, `POST /cart/items`, `DELETE /cart/items/{user_id}/{product_id}`) using `cart/usecases` and `CartItemRepository`. Includes proper validation, error handling, and RESTful standards. Resolves #24.

irvaniamirali committed:
fix(cart): provide proper CartItemRepository instance for dependency injection

irvaniamirali committed:
chore: update uv.lock after dependency changes

irvaniamirali committed:
chore: update project dependencies for cart module changes
- Ensure new packages/hooks are included for linting and formatting

irvaniamirali committed:
feat(cart): implement and standardize all Cart endpoints

- POST /cart/ for bulk item creation using CreateCartFromItems use case
- GET /cart/{user_id} to retrieve a user's cart using ReadCart use case
- DELETE /cart/{user_id} to clear a user's cart
- POST /cart/items to add a single item using AddItemToCart use case
- DELETE /cart/{user_id}/items/{product_id} to remove a single item using RemoveItemFromCart use case
- Follow RESTful conventions, proper JSON request/response formats
- Include error handling for duplicates and missing items

irvaniamirali committed:
chore(cart): include Cart router in main FastAPI app

irvaniamirali committed:
feat(cart): implement AddItemToCart use case for single item addition

irvaniamirali committed:
feat(cart): implement CreateCartFromItems use case for bulk cart creation

irvaniamirali committed:
feat(cart): implement ReadCart use case for retrieving user cart contents

irvaniamirali committed:
feat(cart): implement RemoveItemFromCart use case for deleting single cart item

irvaniamirali committed:
feat(cart): implement CartItemRepository with CRUD and remove methods

irvaniamirali committed:
refactor(cart): define CartItemRepositoryInterface for consistent repository abstraction

babakjahan commented on app/modules/cart/depends.py:
```diff
@@ -3,8 +3,9 @@
 from .repository_interface import CartItemRepositoryInterface
 
 
 def get_cartitem_repository(db_session: AsyncSession = Depends(get_db_session)) -> CartItemRepositoryInterface:
```
db_session: Annotated[AsyncSession, Depends(get_db_session)]
after I talked to you and say you need to follow PEP-593 !
it still has problem

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+        try:
+            self.session.add_all(cart_items)
+            await self.session.commit()
+            for ci in cart_items:
```
use more descriptive names like item or even cart_item.

babakjahan commented on app/modules/cart/repository_interface.py:
```diff
@@ -13,9 +16,17 @@ async def insert(self, cart_item: CartItem) -> CartItem:
 
     @abstractmethod
-    async def update(self, cart_item: CartItem) -> CartItem:
+    async def remove(self, user_id, product_id):
```
return type annotations removed or left unspecified !! why !

babakjahan commented on app/modules/cart/routers.py:
```diff
@@ -32,14 +33,120 @@ async def cart_bulk_create(
+    product_id: int,
+    cartitem_repository: Annotated[CartItemRepositoryInterface, Depends(get_cartitem_repository)],
+) -> SuccessResponse[None]:
+    await RemoveItemFromCart(cartitem_repository).execute(user_id, product_id)
```
if not removed what is gonna happen?

irvaniamirali replied:
I’ve updated the `RemoveItemFromCart` use case to handle failure scenarios explicitly.
- If the item does not exist, `EntityNotFoundException` is raised (→ 404).
- If deletion fails at DB level, `DatabaseOperationException` is raised (→ 500).
- The method returns `None` only when removal is successful.

babakjahan commented on app/modules/cart/usecases/add_item.py:
```diff
@@ -1,22 +1,62 @@
+
+        return self._map_to_read(inserted)
+
+    def _map_to_read(self, cart_item) -> CartItemRead:
```
If CartItemRead mapping depends on async operations or extra DB lookups (e.g., title, sku, price), a sync method may not suffice. need async method

its not good place you leave it in the usecase!

babakjahan commented on pyproject.toml:
```diff
@@ -10,6 +10,7 @@ dependencies = [
     "greenlet>=3.2.1",
     "orjson>=3.10.18",
     "passlib[argon2]>=1.7.4",
+    "pre-commit>=4.2.0",
```
why you added here ! its dev dependecy !

babakjahan commented on app/modules/cart/usecases/add_item.py:
```diff
@@ -1,22 +1,62 @@
+            sku=getattr(cart_item, "sku", ""),
+            price=getattr(cart_item, "price", 0.0),
+            total=getattr(cart_item, "total", 0.0),
+            date_created_gmt=getattr(cart_item, "date_created", None),
```
getattr(cart_item, "date_created", None) maps to date_created_gmt. The naming may cause confusion if the field is not GMT; ensure consistency.

babakjahan commented on app/modules/cart/usecases/add_item.py:
```diff
@@ -1,22 +1,62 @@
         self.cart_item_repository = cart_item_repository
+        self.product_repository = product_repository
 
     async def execute(self, cart_item_create: CartItemCreate) -> CartItemRead | None:
```
### **No type validation for cart_item_create fields before DB access—consider validating input data before usage**

Explanation:
This method receives a cart_item_create parameter (of type CartItemCreate). However, there is no explicit type validation or value checking for the fields of cart_item_create before these fields are used for database access (such as cart_item_create.product_id, cart_item_create.user_id, and cart_item_create.quantity). The code immediately attempts to look up the product, check cart existence, and insert the item using these fields.

Context:

If invalid data (e.g. wrong types, missing fields, or out-of-range values) is passed in, errors may occur farther down (like failed DB calls or exceptions).
Proper validation (e.g., checking field types, required fields, value ranges) should occur before attempting database operations to catch errors early and provide clear feedback.
Where is this missing?
There is no validation block at the start of execute, nor is there any use of Pydantic .validate() or manual checks for the cart_item_create fields. All usage of fields is direct and assumes they are correct.

Suggestion:
Consider adding explicit input validation (either via Pydantic models, custom checks, or other mechanisms) at the start of the method, before any database calls. This will help ensure data integrity and better error handling.

babakjahan commented on app/modules/cart/usecases/read_cart.py:
```diff
@@ -1,6 +1,9 @@
+from typing import List
```
in Python 3.9+, use list[CartItemRead] for consistency

babakjahan commented on app/modules/cart/usecases/read_cart.py:
```diff
@@ -9,14 +12,42 @@ def __init__(self, cart_item_repository: CartItemRepositoryInterface) -> None:
+        if not cart_items:
+            return None
+
+        items_read: List[CartItemRead] = [
```
just use list

babakjahan commented on app/modules/cart/usecases/read_cart.py:
```diff
@@ -9,14 +12,42 @@ def __init__(self, cart_item_repository: CartItemRepositoryInterface) -> None:
 
     async def execute(self, user_id: int) -> CartRead | None:
```
Return Type: Method returns None or CartRead, which is fine, but some codebases prefer always returning a CartRead (possibly with empty items) for consistency.

babakjahan commented on app/modules/cart/usecases/remove_item.py:
```diff
@@ -7,11 +8,19 @@ def __init__(self, cart_item_repository: CartItemRepositoryInterface) -> None:
+        except Exception as e:
+            raise DatabaseOperationException(operation="delete", message=str(e))
+
+        return None
```
Unnecessary Return Statement: The execute method returns None explicitly, but as it’s an async function returning nothing, this is redundant.

babakjahan commented on app/modules/cart/usecases/remove_item.py:
```diff
@@ -7,11 +8,19 @@ def __init__(self, cart_item_repository: CartItemRepositoryInterface) -> None:
         """
 
-        pass
+        item = await self.cart_item_repository.get_by_user_and_product(user_id, product_id)
```
The get_by_user_and_product and remove methods may throw repository-specific exceptions not handled separately.

babakjahan commented on app/modules/cart/usecases/create_cart.py:
```diff
@@ -1,20 +1,104 @@
+from typing import List, Tuple
```
use list

babakjahan commented on app/modules/cart/usecases/create_cart.py:
```diff
@@ -1,20 +1,104 @@
+        ]
+
+        try:
+            inserted_items = await self.cart_item_repository.bulk_insert(cart_items)
```
No Transaction Handling: Bulk DB operations should use transactions to prevent partial inserts on errors.

babakjahan commented on app/modules/cart/usecases/create_cart.py:
```diff
@@ -1,20 +1,104 @@
+        result = await self.cart_item_repository.session.execute(stmt)
+        existing_pairs = set(result.fetchall())
+
+        if existing_pairs:
```
Error on First Duplicate Only: Only the first duplicate pair is reported, potentially missing others.

babakjahan commented on app/modules/cart/usecases/create_cart.py:
```diff
@@ -1,20 +1,104 @@
+            dup_user_id, dup_product_id = next(iter(existing_pairs))
+            raise DuplicateEntryException(field="product_id", value=str(dup_product_id))
+
+        cart_items = [
```
 Some variables like cart_items, inserted_items, items_read in lines 73, 81, 89 are missing explicit type annotations.

irvaniamirali committed:
refactor(deps): migrate db_session injection to PEP-593 Annotated style

irvaniamirali committed:
refactor(cart): use more descriptive variable names for cart items

irvaniamirali committed:
refactor(cart): add explicit None return type to remove method

irvaniamirali committed:
refactor(cart): improve exception handling in RemoveItemFromCart use case

irvaniamirali committed:
refactor(types): use built-in list and tuple instead of typing.List/Tuple

irvaniamirali committed:
fix(cart): wrap bulk_insert in transaction with exception handling

irvaniamirali committed:
refactor(cart): rename loop variable 'ci' to 'cart_item' for clarity

irvaniamirali committed:
fix(cart): report all duplicate items in CreateCartFromItems use case

irvaniamirali committed:
refactor(cart): improve code readability in CreateCartFromItems

irvaniamirali committed:
refactor(cart): inline CartItemRead mapping in AddItemToCart

irvaniamirali committed:
refactor(cart): use built-in list instead of typing.List in ReadCart

irvaniamirali committed:
refactor(cart): add CartItem type hint for cart_item instance

irvaniamirali committed:
fix(pyproject): remove accidental pre-commit dependency

irvaniamirali committed:
refactor(cart): apply Null Object Pattern in ReadCart usecase

- Always return CartRead instead of None
- Return CartRead with empty items list when user has no cart items
- Simplifies client code and avoids None checks

irvaniamirali committed:
chore(db): standardize database operation names across repository methods
